compiler/rustc_middle/src/mir/consts.rs

```diff
@@ -288,7 +288,16 @@ impl<'tcx> Const<'tcx> {
         tcx: TyCtxt<'tcx>,
         param_env: ty::ParamEnv<'tcx>,
     ) -> Option<ScalarInt> {
-        self.try_eval_scalar(tcx, param_env)?.try_to_int().ok()
+        match self {
+            // If the constant is already evaluated, we shortcut here.
+            Const::Ty(c) if let ty::ConstKind::Value(valtree) = c.kind() => {
+                valtree.try_to_scalar_int()
```

Comment:
Ah, there's the `valtree_to_const_val`. But then we should really have this as a fast-path in `try_eval_scalar` and it should apply on all `Const::Ty`, not just for things that are already valtrees.

https://github.com/rust-lang/rust/pull/116457

Reply:
I considered it, but the optimization is only correct if we know the normal path will only find Valtree::Leaf I think. We talked about it here https://github.com/rust-lang/rust/pull/116281#discussion_r1342189369

Reply:
I mostly felt out of my depth on this ngl

Reply:
Such discussion should definitely have resulted in a comment in the code...

Reply:
This is still a change in behavior. Previously, `try_eval_scalar_int` would have worked on struct newtypes around integers; now it might return `None` since those are branch valtrees (under our current representation).

Also, previously if this accidentally got called on a reference it would have returned `None`, but now when called on `&i32` it will return the integer value. That seems bad.

Reply:
Oh that's not good at all. I did run all tests once with an assert checking that it's only called on numeric types fwiw. Also the speedup was ~1.5% on unicode-normalization

Reply:
>  I did run all tests once with an assert checking that it's only called on numeric types fwiw.

That doesn't help guard against new uses introduced tomorow.

Reply:
Ah yeah that's much better